corehq/apps/hqcase/api/updates.py

```diff
@@ -275,3 +300,82 @@ def _submit_case_blocks(case_blocks, domain, user, device_id):
         device_id=device_id,
         max_wait=15
     )
+
+
+def _get_owner_ids(domain, data, case_db):
+    owner_ids = []
+    case_ids = []
+    for data_item in data:
+        if data_item.is_new_case:
+            owner_ids.append(data_item.owner_id)
```

Comment:
Shouldn't this append the owner ID also for existing cases?  Otherwise it'd let restricted users reassign existing cases they have access to to owners they don't

Reply:
The `owner_id` is optional for updating existing cases, so we can't rely on appending it since it might not be there. Either a `case_id` or `external_id` is required however, so we know these will always exist for case updates. These are used to get the `owner_id` of the case using the query further down, which we then do the necessary permission checking on. 

Since we ultimately end up checking that the user has access to the `owner_id` of all the `case_ids` we append, they shouldn't be able to update/reassign cases they don't have access to.

Reply:
I'm thinking of the reverse sort of scenario - let's consider that a user does have access to a case and they submit an update assigning it to an owner they cannot access.  I _think_ this should be prohibited (though that's not totally clear-cut to me).

In that case, `is_new_case` will be `False`, and `owner_id` will be set.  We'll look up the existing owner ID by the `case_id`, and the user will have access to it, but we don't validate that the user has access to the new `owner_id`.

Reply:
Ah I see what you mean. I wonder if this is something we should prohibit though, and my reasoning is regarding referrals. For example:

Let's say Dr. Alice works in hospital A and B and Dr. John works in hospital B. Should Dr. John be able to refer a patient to Dr. Alice in hospital A, even though he doesn't have access to it? Perhaps in this case, there should be an intermediary that has access to both hospitals that needs to do the referral transfer of the patient from Dr. John to Dr. Alice? I'm not entirely sure how restrictive we should be on this. What are your thoughts on this @esoergel?

Reply:
Yeah, I was thinking about referrals as well, I agree that that does sound plausibly useful.  However I'm not positive, so I'm inclined to err on the side of being restrictive.  Furthermore, this code already prevents users from _creating_ cases assigned to owners they can't access, so it seems consistent to also prevent users from reassigning existing cases to owners they can't access (otherwise it'd be easy to circumvent the other restriction by creating a case then reassigning it).

Reply:
That is a good point. I'll create a separate PR to address this, where I'll check if an `owner_id` is present on a case update. If it is, then we can use that, along with the `case_id`. If not, then we'll only use the latter.

Reply:
thanks!